pos_blstm/blstm.py

```diff
@@ -0,0 +1,206 @@
+        print('train acc:',sum(np.array(sentence) == np.array(correct))/float(len(sentence)))
+
+
+        logprobs = lstm_predict(weights, X_test[:,:num_sentences,:], cap_test[:,:num_sentences,:])
+        sentence = []
+        correct = []
+        for i in xrange(num_sentences):
+            for j in xrange(test_mask[i]):
+                sentence.append(one_hot[np.argmax(np.exp(logprobs)[:,i,:][j])])
+                correct.append(one_hot[np.argmax(Y_test[j,i,:])])
+        print('valid acc:',sum(np.array(sentence) == np.array(correct))/float(len(sentence)))
+        # print("Training text                         Predicted text")
+        # logprobs = np.asarray(lstm_predict(weights, train_inputs))
+        # for t in range(logprobs.shape[1]):
+        #     training_text  = one_hot_to_string(train_inputs[:,t,:])
+        #     predicted_text = one_hot_to_string(logprobs[:,t,:])
+        #     print(training_text.replace('\n', ' ') + "|" +
+        #           predicted_text.replace('\n', ' '))
+
+    def training_loss(params, iter):
+        samples = 50
+        start_index = npr.randint(0,X_train.shape[1]-samples)
+        # start_index = 0
+        # samples = 50
+        return -lstm_log_likelihood(params, X_train[:,start_index:start_index+samples,:], Y_train[:,start_index:start_index+samples,:], train_mask[start_index:start_index+samples], cap_train[:,start_index:start_index+samples,:])
+
+    def callback(weights, iter, gradient):
+        if iter % 10 == 0:
+            print("Iteration", iter, "Train loss:", training_loss(weights, 0))
+            print_training_prediction(weights)
+        if iter % 250 == 0 and save:
+            np.savez('tmp/trained_weights.npz', **weights)
+
+    # Build gradient of loss function using autograd.
+    training_loss_grad = grad(training_loss)
+
+    print("Training LSTM...")
+    trained_params = adam(training_loss_grad, init_params, step_size=0.01,
+                          num_iters=2000, callback=callback)
+
+
+    num_sentences = 10
+    logprobs = lstm_predict(trained_params, X_train[:,:num_sentences,:], cap_train[:,:num_sentences,:])
+    sentence = []
+    correct = []
+    for i in xrange(num_sentences):
+        for j in xrange(train_mask[i]):
+            sentence.append(one_hot[np.argmax(np.exp(logprobs)[:,i,:][j])])
+            correct.append(one_hot[np.argmax(Y_train[j,i,:])])
+    print(sum(np.array(sentence) == np.array(correct))/float(len(sentence)))
+
+    logprobs = lstm_predict(trained_params, X_test[:,:num_sentences,:], cap_test[:,:num_sentences,:])
+    sentence = []
+    correct = []
+    for i in xrange(num_sentences):
+        for j in xrange(test_mask[i]):
+            sentence.append(one_hot[np.argmax(np.exp(logprobs)[:,i,:][j])])
+            correct.append(one_hot[np.argmax(Y_test[j,i,:])])
+    print(sum(np.array(sentence) == np.array(correct))/float(len(sentence)))
+
+    if save:
+        np.savez('tmp/trained_weights.npz', **trained_params)
```

Comment:
Does it currently work? You should compare it with the standard example  from the LSTM file in github. This should do no worse. (https://github.com/HIPS/autograd/blob/master/examples/lstm.py)